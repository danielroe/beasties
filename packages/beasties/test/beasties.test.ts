/**
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

import { describe, expect, it, vi } from 'vitest'

import Beasties from '../src/index'

const fixtureDir = fileURLToPath(new URL('./src', import.meta.url))

function trim(s: TemplateStringsArray) {
  return s[0]!
    .trim()
    .replace(new RegExp(`^${s[0]!.match(/^( {2}|\t)+/m)![0]}`, 'gm'), '')
}

describe('beasties', () => {
  it('basic Usage', async () => {
    const beasties = new Beasties({
      reduceInlineStyles: false,
      path: '/',
    })
    const assets: Record<string, string> = {
      '/style.css': trim`
        h1 { color: blue; }
        h2.unused { color: red; }
        p { color: purple; }
        p.unused { color: orange; }
      `,
    }
    beasties.readFile = filename => assets[filename.replace(/^\w:/, '').replace(/\\/g, '/')]!
    const result = await beasties.process(trim`
      <html>
        <head>
          <link rel="stylesheet" href="/style.css">
        </head>
        <body>
          <h1>Hello World!</h1>
          <p>This is a paragraph</p>
        </body>
      </html>
    `)
    expect(result).toMatch('<style>h1{color:blue}p{color:purple}</style>')
    expect(result).toMatch('<link rel="stylesheet" href="/style.css">')
    expect(result).toMatchSnapshot()
  })

  it('works with an html snippet', async () => {
    const beasties = new Beasties()
    const result = await beasties.process(trim`
      <style>
        .red { color: red }
        .blue { color: blue }
      </style>
      <div class="blue">I'm Blue</div>
    `)
    expect(result).toMatchInlineSnapshot(`
      "<style>.blue{color:blue}</style>
      <div class="blue">I'm Blue</div>"
    `)
  })

  it('run on HTML file', async () => {
    const beasties = new Beasties({
      reduceInlineStyles: false,
      path: fixtureDir,
    })

    const html = fs.readFileSync(path.join(fixtureDir, 'index.html'), 'utf-8')

    const result = await beasties.process(html)
    expect(result).toMatchSnapshot()
  })

  it('does not encode HTML', async () => {
    const beasties = new Beasties({
      reduceInlineStyles: false,
      path: '/',
    })
    const assets: Record<string, string> = {
      '/style.css': trim`
        h1 { color: blue; }
      `,
    }
    beasties.readFile = filename => assets[filename.replace(/^\w:/, '').replace(/\\/g, '/')]!
    const result = await beasties.process(trim`
      <html>
        <head>
          <title>$title</title>
          <link rel="stylesheet" href="/style.css">
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    `)
    expect(result).toMatch('<style>h1{color:blue}</style>')
    expect(result).toMatch('<link rel="stylesheet" href="/style.css">')
    expect(result).toMatch('<title>$title</title>')
  })

  it('should keep existing link tag attributes in the noscript link', async () => {
    const beasties = new Beasties({
      reduceInlineStyles: false,
      path: '/',
      preload: 'media',
    })
    const assets: Record<string, string> = {
      '/style.css': trim`
        h1 { color: blue; }
      `,
    }
    beasties.readFile = filename => assets[filename.replace(/^\w:/, '').replace(/\\/g, '/')]!
    const result = await beasties.process(trim`
      <html>
        <head>
          <title>$title</title>
          <link rel="stylesheet" href="/style.css" crossorigin="anonymous" integrity="sha384-j1GsrLo96tLqzfCY+">
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    `)

    expect(result).toMatch('<style>h1{color:blue}</style>')
    expect(result).toMatch(
      `<link rel="stylesheet" href="/style.css" crossorigin="anonymous" integrity="sha384-j1GsrLo96tLqzfCY+" media="print" onload="this.media='all'">`,
    )
    expect(result).toMatchSnapshot()
  })

  it('should keep existing link tag attributes', async () => {
    const beasties = new Beasties({
      reduceInlineStyles: false,
      path: '/',
    })
    const assets: Record<string, string> = {
      '/style.css': trim`
        h1 { color: blue; }
      `,
    }
    beasties.readFile = filename => assets[filename.replace(/^\w:/, '').replace(/\\/g, '/')]!
    const result = await beasties.process(trim`
      <html>
        <head>
          <title>$title</title>
          <link rel="stylesheet" href="/style.css" crossorigin="anonymous" integrity="sha384-j1GsrLo96tLqzfCY+">
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    `)

    expect(result).toMatch('<style>h1{color:blue}</style>')
    expect(result).toMatch(
      `<link rel="stylesheet" href="/style.css" crossorigin="anonymous" integrity="sha384-j1GsrLo96tLqzfCY+">`,
    )
    expect(result).toMatchSnapshot()
  })

  it('does not decode entities in HTML document', async () => {
    const beasties = new Beasties({
      path: '/',
    })
    const assets: Record<string, string> = {
      '/style.css': trim`
        h1 { color: blue; }
        h2.unused { color: red; }
        p { color: purple; }
        p.unused { color: orange; }
      `,
    }
    beasties.readFile = filename => assets[filename.replace(/^\w:/, '').replace(/\\/g, '/')]!
    const result = await beasties.process(trim`
      <html>
        <body>
          &lt;h1&gt;Hello World!&lt;/h1&gt;
        </body>
      </html>
    `)
    expect(result).toMatch('&lt;h1&gt;Hello World!&lt;/h1&gt;')
  })

  it('prevent injection via media attr', async () => {
    const beasties = new Beasties({
      reduceInlineStyles: false,
      path: fixtureDir,
      preload: 'media',
    })

    const html = fs.readFileSync(path.join(fixtureDir, 'media-validation.html'), 'utf-8')

    const result = await beasties.process(html)
    expect(result).toContain(
      '<noscript><link rel="stylesheet" href="styles2.css" media="screen and (min-width: 480px)"></noscript>',
    )
    expect(result).toMatchSnapshot()
  })

  it('skip invalid path', async () => {
    const consoleSpy = vi.spyOn(console, 'warn')

    const beasties = new Beasties({
      reduceInlineStyles: false,
      path: fixtureDir,
    })

    const html = fs.readFileSync(path.join(fixtureDir, 'subpath-validation.html'), 'utf-8')

    const result = await beasties.process(html)
    expect(consoleSpy).not.toHaveBeenCalledWith(
      expect.stringContaining('Unable to locate stylesheet'),
    )
    expect(result).toMatchSnapshot()
  })

  it('should not load stylesheets outside of the base path', async () => {
    const beasties = new Beasties({ path: '/var/www' })
    vi.spyOn(beasties, 'readFile')
    await beasties.process(`
        <html>
            <head>
                <link rel=stylesheet href=/file.css>
                <link rel=stylesheet href=/../../../company-secrets/secret.css>
            </head>
            <body></body>
        </html>
    `)
    expect(beasties.readFile).toHaveBeenCalledWith(path.resolve('/var/www/file.css'))
    expect(beasties.readFile).not.toHaveBeenCalledWith(
      '/company-secrets/secret.css',
    )
  })
})
