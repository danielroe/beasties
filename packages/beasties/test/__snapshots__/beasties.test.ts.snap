// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`beasties > basic Usage 1`] = `
"<html data-beasties-container>
  <head>
    <style>h1{color:blue}p{color:purple}</style><link rel="preload" href="/style.css" as="style">
  </head>
  <body>
    <h1>Hello World!</h1>
    <p>This is a paragraph</p>
  <link rel="stylesheet" href="/style.css"></body>
</html>"
`;

exports[`beasties > prevent injection via media attr 1`] = `
"<html>
<link>
<title>Testing</title>
<style>h1{color:blue}p{color:purple}.contents{padding:50px;text-align:center}.input-field{padding:10px}.custom-element::part(tab){color:#0c0dcc;border-bottom:transparent solid 2px}.custom-element::part(tab):hover:active{background-color:#0c0d33;color:#ffffff}.custom-element::part(tab):focus{box-shadow:0 0 0 1px #0a84ff inset, 0 0 0 1px #0a84ff,
    0 0 0 4px rgba(10, 132, 255, 0.3)}div:is(:hover, .active){color:#000}div:is(.selected, :hover){color:#fff}body{height:100%}</style><link rel="stylesheet" href="styles.css" media="print" onload="this.media='all'"><noscript><link rel="stylesheet" href="styles.css" media="alert(1)"></noscript>

<link rel="stylesheet" href="styles2.css" media="print" onload="this.media='screen and (min-width: 480px)'"><noscript><link rel="stylesheet" href="styles2.css" media="screen and (min-width: 480px)"></noscript>



<body>
  <div class="container">
    <header>
      <div class="banner">Lorem ipsum dolor sit amet</div>
    </header>
  </div>
  <div class data-beasties-container>
    <h1>Hello World!</h1>
    <p>This is a paragraph</p>
    <input class="input-field">
    <div class="contents"></div>
    <div class="selected active"></div>
  </div>
  <footer class></footer>
</body>

</html>
"
`;

exports[`beasties > run on HTML file 1`] = `
"<html>
  <head>
    <title>Testing</title>
    <style>h1{color:blue}p{color:purple}.contents{padding:50px;text-align:center}.input-field{padding:10px}.custom-element::part(tab){color:#0c0dcc;border-bottom:transparent solid 2px}.custom-element::part(tab):hover:active{background-color:#0c0d33;color:#ffffff}.custom-element::part(tab):focus{box-shadow:0 0 0 1px #0a84ff inset, 0 0 0 1px #0a84ff,
    0 0 0 4px rgba(10, 132, 255, 0.3)}div:is(:hover, .active){color:#000}div:is(.selected, :hover){color:#fff}</style><link rel="preload" href="styles.css" as="style">
  </head>
  <body>
    <div class="container">
      <header>
        <div class="banner">Lorem ipsum dolor sit amet</div>
      </header>
    </div>
    <div class data-beasties-container>
      <h1>Hello World!</h1>
      <p>This is a paragraph</p>
      <input class="input-field">
      <div class="contents"></div>
      <div class="selected active"></div>
    </div>
    <footer class></footer>
  <link rel="stylesheet" href="styles.css"></body>
</html>
"
`;

exports[`beasties > should keep existing link tag attributes 1`] = `
"<html data-beasties-container>
  <head>
    <title>$title</title>
    <style>h1{color:blue}</style><link rel="preload" href="/style.css" crossorigin="anonymous" integrity="sha384-j1GsrLo96tLqzfCY+" as="style">
  </head>
  <body>
    <h1>Hello World!</h1>
  <link rel="stylesheet" href="/style.css" crossorigin="anonymous" integrity="sha384-j1GsrLo96tLqzfCY+"></body>
</html>"
`;

exports[`beasties > should keep existing link tag attributes in the noscript link 1`] = `
"<html data-beasties-container>
  <head>
    <title>$title</title>
    <style>h1{color:blue}</style><link rel="stylesheet" href="/style.css" crossorigin="anonymous" integrity="sha384-j1GsrLo96tLqzfCY+" media="print" onload="this.media='all'"><noscript><link rel="stylesheet" href="/style.css" crossorigin="anonymous" integrity="sha384-j1GsrLo96tLqzfCY+"></noscript>
  </head>
  <body>
    <h1>Hello World!</h1>
  </body>
</html>"
`;

exports[`beasties > skip invalid path 1`] = `
"<html>
<link>
<title>Testing</title>
<style>h1{color:blue}p{color:purple}.contents{padding:50px;text-align:center}.input-field{padding:10px}.custom-element::part(tab){color:#0c0dcc;border-bottom:transparent solid 2px}.custom-element::part(tab):hover:active{background-color:#0c0d33;color:#ffffff}.custom-element::part(tab):focus{box-shadow:0 0 0 1px #0a84ff inset, 0 0 0 1px #0a84ff,
    0 0 0 4px rgba(10, 132, 255, 0.3)}div:is(:hover, .active){color:#000}div:is(.selected, :hover){color:#fff}</style><link rel="preload" href="styles.css" as="style">

<link rel="stylesheet" href="../../../../../style3.css">



<body>
  <div class="container">
    <header>
      <div class="banner">Lorem ipsum dolor sit amet</div>
    </header>
  </div>
  <div class data-beasties-container>
    <h1>Hello World!</h1>
    <p>This is a paragraph</p>
    <input class="input-field">
    <div class="contents"></div>
    <div class="selected active"></div>
  </div>
  <footer class></footer>
<link rel="stylesheet" href="styles.css"></body>

</html>
"
`;

exports[`beasties > works with pseudo classes and elements 1`] = `
"<html data-beasties-container>
  <head>
    <style>h1{color:blue}h1:has(+ p){margin-bottom:0}p{color:purple}p:only-child{color:fuchsia}input:where(:not([readonly])):where(:active, :focus, :focus-visible, [data-focused]){color:blue}</style><link rel="preload" href="/style.css" as="style">
  </head>
  <body>
    <h1>Hello World!</h1>
    <p>This is a paragraph</p>
    <input type="text">
  <link rel="stylesheet" href="/style.css"></body>
</html>"
`;

exports[`preload modes > should handle "false" preload mode correctly 1`] = `
"
      <html data-beasties-container>
        <head>
          <style>h1{color:blue}</style><link rel="stylesheet" href="/style.css">
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    "
`;

exports[`preload modes > should use "js" preload mode correctly 1`] = `
"
      <html data-beasties-container>
        <head>
          <style>h1{color:blue}</style><link rel="preload" href="/style.css" as="style"><noscript><link rel="stylesheet" href="/style.css"></noscript><script data-href="/style.css" data-media="all">function $loadcss(u,m,l){(l=document.createElement('link')).rel='stylesheet';l.href=u;document.head.appendChild(l)}$loadcss(document.currentScript.dataset.href,document.currentScript.dataset.media)</script>
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    "
`;

exports[`preload modes > should use "media" preload mode correctly 1`] = `
"
      <html data-beasties-container>
        <head>
          <style>h1{color:blue}</style><link rel="stylesheet" href="/style.css" media="print" onload="this.media='all'"><noscript><link rel="stylesheet" href="/style.css"></noscript>
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    "
`;

exports[`preload modes > should use "swap" preload mode correctly 1`] = `
"
      <html data-beasties-container>
        <head>
          <style>h1{color:blue}</style><link rel="preload" href="/style.css" onload="this.rel='stylesheet'" as="style"><noscript><link rel="stylesheet" href="/style.css"></noscript>
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    "
`;

exports[`preload modes > should use "swap-high" preload mode correctly 1`] = `
"
      <html data-beasties-container>
        <head>
          <style>h1{color:blue}</style><link rel="alternate stylesheet preload" href="/style.css" title="styles" onload="this.title='';this.rel='stylesheet'"><noscript><link rel="stylesheet" href="/style.css"></noscript>
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    "
`;

exports[`preload modes > should use "swap-low" preload mode correctly 1`] = `
"
      <html data-beasties-container>
        <head>
          <style>h1{color:blue}</style><link rel="alternate stylesheet" href="/style.css" title="styles" onload="this.title='';this.rel='stylesheet'"><noscript><link rel="stylesheet" href="/style.css"></noscript>
        </head>
        <body>
          <h1>Hello World!</h1>
        </body>
      </html>
    "
`;
